src/Castle.Windsor.Extensions.DependencyInjection/Scope/ExtensionContainerScope.cs

```diff
@@ -63,20 +63,21 @@ public void Dispose()
 
 		public Burden GetCachedInstance(ComponentModel model, ScopedInstanceActivationCallback createInstance)
 		{
-			if(model.Configuration.Attributes.Get(TransientMarker) == Boolean.TrueString )
-			{
-				var burden = createInstance((_) => {});
-				scopeCache[burden] = burden;
-				return burden;
-			}
-			else
+			lock (scopeCache)
 			{
-				var burden = scopeCache[model];
-				if (burden == null)
+				Burden burden = null;
+				if(model.Configuration.Attributes.Get(TransientMarker) != bool.TrueString )
 				{
-					scopeCache[model] = burden = createInstance((_) => {});
+					burden = scopeCache[model];
 				}
-
+				else
+				{
+					burden = createInstance((_) => {});
+					scopeCache[burden] = burden;
+				}
+				if (burden != null) return burden;
+				burden = createInstance((_) => {});
+				scopeCache[model] = burden;
```

Comment:
There is some duplication in this method, maybe go with this:

```csharp
		public Burden GetCachedInstance(ComponentModel model, ScopedInstanceActivationCallback createInstance)
		{
			lock (scopeCache)
			{
				Burden burden = null;

				if (model.Configuration.Attributes.Get(TransientMarker) != bool.TrueString)
				{
					burden = scopeCache[model];
				}

				if (burden == null)
				{
					burden = createInstance((_) => {});
					scopeCache[burden] = burden;
				}

				return burden;
			}
		}
```

Reply:
Maybe this would make it clearer what is happening:

```csharp
		public Burden GetCachedInstance(ComponentModel model, ScopedInstanceActivationCallback createInstance)
		{
			lock (scopeCache)
			{
				Burden burden = null;

				// Add transient's burden to scope so it gets released
				if (model.Configuration.Attributes.Get(TransientMarker) == bool.TrueString)
				{
					burden = createInstance((_) => {});
					scopeCache[burden] = burden;
					return burden;
				}

				burden = scopeCache[model];
				if (burden == null)
				{
					burden = createInstance((_) => {});
					scopeCache[model] = burden;
				}
				return burden;
			}
		}
```